File: app/src/main/java/com/android/ashwiask/tvmaze/home/ShowsAdapter.java
package com.android.ashwiask.tvmaze.home;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.content.res.AppCompatResources;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.android.ashwiask.tvmaze.R;
import com.android.ashwiask.tvmaze.databinding.ShowListItemBinding;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.bumptech.glide.request.RequestOptions;

import java.util.Collections;
import java.util.List;

/**
 * @author Ashwini Kumar.
 */

public class ShowsAdapter extends RecyclerView.Adapter<ShowsAdapter.ShowHolder> {
    private static final String ORIGINAL_IMAGE = "original";
    private List<Show> shows;
    private final Callback callback;

    ShowsAdapter(Callback callback) {
        this.callback = callback;
        shows = Collections.emptyList();
    }

    public void updateList(List<Show> newList) {
        ShowDiffUtilCallback showDiffUtilCallback = new ShowDiffUtilCallback(shows, newList);
        final DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(showDiffUtilCallback);
        shows = newList;
        diffResult.dispatchUpdatesTo(this);
    }

    @NonNull
    @Override
    public ShowHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        ShowListItemBinding showListItemBinding = ShowListItemBinding
                .inflate(layoutInflater, parent, false);
        ShowHolder showHolder = new ShowHolder(showListItemBinding);
        showHolder.binding.favorite.setOnClickListener(v ->
                onFavouriteIconClicked(showHolder.getAdapterPosition()));
        return showHolder;
    }

    private void onFavouriteIconClicked(int position) {
        if (position != RecyclerView.NO_POSITION) {
            Show show = shows.get(position);
            show = show.toBuilder().isFavorite(!show.isFavorite())
                    .build();
            shows.set(position, show);
            notifyItemChanged(position);
            callback.onFavoriteClicked(show);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull ShowHolder holder, int position) {
        Show show = shows.get(position);
        configureImage(holder.binding.showImage, show);
        configureFavoriteIcon(holder.binding.favorite, show.isFavorite());
    }

    @Override
    public void onBindViewHolder(@NonNull ShowHolder holder, int position,
                                 @NonNull List<Object> payloads) {
        if (!payloads.isEmpty()) {
            Bundle bundle = (Bundle) payloads.get(0);
            boolean isFavorite = bundle.getBoolean("IS_FAVORITE");
            configureFavoriteIcon(holder.binding.favorite, isFavorite);
        } else {
            onBindViewHolder(holder, position);
        }
    }

    private void configureFavoriteIcon(ImageView favoriteIcon, boolean favorite) {
        if (favorite) {
            Drawable favoriteDrawable = AppCompatResources
                    .getDrawable(favoriteIcon.getContext(), R.drawable.favorite);
            favoriteIcon.setImageDrawable(favoriteDrawable);
        } else {
            Drawable unFavoriteDrawable = AppCompatResources
                    .getDrawable(favoriteIcon.getContext(), R.drawable.favorite_border);
            favoriteIcon.setImageDrawable(unFavoriteDrawable);
        }
    }

    private void configureImage(ImageView showImage, Show show) {
        if (show.image() != null) {
            Glide.with(showImage.getContext()).load(show.image().get(ORIGINAL_IMAGE))
                    .apply(RequestOptions.placeholderOf(R.color.grey))
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(showImage);
        }
    }

    @Override
    public int getItemCount() {
        return shows.size();
    }

    public List<Show> getShows() {
        return shows;
    }

    static class ShowHolder extends RecyclerView.ViewHolder {
        private ShowListItemBinding binding;

        ShowHolder(ShowListItemBinding itemBinding) {
            super(itemBinding.getRoot());
            this.binding = itemBinding;
        }
    }

    public interface Callback {
        void onFavoriteClicked(Show show);
    }
}

File: app/src/main/java/com/android/ashwiask/tvmaze/home/ShowsAdapter.kt
package com.android.ashwiask.tvmaze.home

import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.content.res.AppCompatResources
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.android.ashwiask.tvmaze.R
import com.android.ashwiask.tvmaze.databinding.ShowListItemBinding
import com.android.ashwiask.tvmaze.network.home.Show
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestOptions

const val IS_FAVORITE = "IS_FAVORITE"

class ShowsAdapter constructor(
    private val callback: Callback
) : RecyclerView.Adapter<ShowsAdapter.ShowHolder>() {
    private var episodeViewDataList: MutableList<HomeViewData.EpisodeViewData> = mutableListOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ShowHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val showListItemBinding = ShowListItemBinding
            .inflate(layoutInflater, parent, false)
        val showHolder = ShowHolder(showListItemBinding)
        showHolder.binding.showFavoriteIcon = true
        showHolder.binding.favorite.setOnClickListener { onFavouriteIconClicked(showHolder.adapterPosition) }
        return showHolder
    }

    fun updateList(newList: MutableList<HomeViewData.EpisodeViewData>) {
        val showDiffUtilCallback = ShowDiffUtilCallback(episodeViewDataList, newList)
        val diffResult = DiffUtil.calculateDiff(showDiffUtilCallback)
        episodeViewDataList = newList
        diffResult.dispatchUpdatesTo(this)
    }

    private fun onFavouriteIconClicked(position: Int) {
        if (position != RecyclerView.NO_POSITION) {
            val episodeViewData = episodeViewDataList[position]
            val isFavorite = episodeViewData.showViewData.isFavoriteShow
            val updatedShowViewData = episodeViewData.showViewData.copy(isFavoriteShow = !isFavorite)
            val updatedEpisodeViewData = episodeViewData.copy(showViewData = updatedShowViewData)
            episodeViewDataList[position] = updatedEpisodeViewData
            notifyItemChanged(position)
            callback.onFavoriteClicked(episodeViewData.showViewData)
        }
    }

    override fun onBindViewHolder(holder: ShowHolder, position: Int) {
        val episodeViewData = episodeViewDataList[position]
        configureImage(holder.binding.showImage, episodeViewData.showViewData.show)
        configureFavoriteIcon(holder.binding.favorite, episodeViewData.showViewData.isFavoriteShow)
    }

    override fun onBindViewHolder(
        holder: ShowHolder, position: Int,
        payloads: List<Any>
    ) {
        if (payloads.isNotEmpty()) {
            val bundle = payloads[0] as Bundle
            val isFavorite = bundle.getBoolean(IS_FAVORITE)
            configureFavoriteIcon(holder.binding.favorite, isFavorite)
        } else {
            onBindViewHolder(holder, position)
        }
    }

    private fun configureFavoriteIcon(favoriteIcon: ImageView, favorite: Boolean) {
        if (favorite) {
            val favoriteDrawable = AppCompatResources
                .getDrawable(favoriteIcon.context, R.drawable.favorite)
            favoriteIcon.setImageDrawable(favoriteDrawable)
        } else {
            val unFavoriteDrawable = AppCompatResources
                .getDrawable(favoriteIcon.context, R.drawable.favorite_border)
            favoriteIcon.setImageDrawable(unFavoriteDrawable)
        }
    }

    private fun configureImage(showImage: ImageView, show: Show) {
        if (show.image != null) {
            Glide.with(showImage.context).load(show.image[ORIGINAL_IMAGE])
                .apply(RequestOptions.placeholderOf(R.color.grey))
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(showImage)
        }
    }

    override fun getItemCount(): Int {
        return episodeViewDataList.size
    }

    class ShowHolder(val binding: ShowListItemBinding) : RecyclerView.ViewHolder(binding.root)

    interface Callback {
        fun onFavoriteClicked(showViewData: HomeViewData.ShowViewData)
    }

    companion object {
        private const val ORIGINAL_IMAGE = "original"
    }
}
